core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java

```diff
@@ -284,6 +286,18 @@ public void messageReceived(PingRequest request, TransportChannel channel) throw
                     + clusterName + "]");
             }
 
+            final DiscoveryNode currentMasterNode = clusterService.state().nodes().getMasterNode();
+            if (currentMasterNode != null
+                    && request.masterNode != null
+                    && currentMasterNode.equals(request.masterNode) == false) {
+                // this node has a different node as master than the one that sent this ping message, so notify
+                // the sending node that this node does not believe it to be the master
+                // TODO: should we throw a specialized exception type here that can be caught by the
```

Comment:
that retry sounds like a  bug. looking at the places where we throw `IllegalStateException` we should never retry. I vote for changing that (and only retry on timeouts + unknown exceptions (with a warning))